### ISSUE TITLE
Can't seem to apply prop constraints via ReactClass<>

### ISSUE THREAD

**Issue by glenjamin** [facebook/flow#1777]
I'm trying to specify the type of components I'm importing from a module, have reduced down to this simple case:

This correctly fails the type check due to missing prop

``` jsx
/* @flow */
/* eslint-env browser */
import React from "react";
import ReactDOM from "react-dom";

class App extends React.Component {
  props: { required: mixed };
}

ReactDOM.render(
  <App />,
  document.getElementById("app")
);

// temp.js:7
//   7:   props: { required: mixed };
//               ^^^^^^^^^^^^^^^^^^^ property `required`. Property not found in
//  11:   <App />,
//        ^^^^^^^ props of React element `App`
```

This next example simulates applying a type to a foreign component, it does not error:

``` jsx
/* @flow */
/* eslint-env browser */
import React from "react";
import ReactDOM from "react-dom";

class App extends React.Component {
  props: { required: mixed };
}
const App2: any = App;
const App3: ReactClass<{required: mixed}> = App2;

ReactDOM.render(
  <App3 />,
  document.getElementById("app")
);
```

**Comment by nanot1m** [comment]
Same problem. Seems flow do not checking ReactClass at all

**Comment by avikchaudhuri** [comment]
`ReactClass` is pretty broken, yes. Will fix.

**Comment by ForbesLindesay** [comment]
Is there any chance this could be fixed soon?  Very real bugs are being ignored by `ReactClass` not actually checking anything.

**Comment by vkurchatkin** [comment]
@ForbesLindesay what is your use case?

**Comment by ForbesLindesay** [comment]
The use case is for higher order components:

```js
type PropsProvidedByConnect = {
  // ...some list of props...
};
type PropsRequiredByConnect = {
  // ...some list of props...
};
function connect<ComponentProps>(
  Component: ReactClass<ComponentProps & PropsProvidedByConnect>,
): ReactClass<ComponentProps & PropsRequiredByConnect> {
  // return a wrapper component that passes through any un-recognised properties,
  // and also adds some extra properties and uses some properties directly.
}
```

I use this kind of pattern pretty frequently for data fetching.

**Comment by asolove** [comment]
This seems to be fixed as of Flow 0.53, which changed the ways prop types are tracked on Component classes. If I make minor changes to the original code above, it gives a type error as expected:


```js
/* @flow */
/* eslint-env browser */
import * as React from "react";
import * as ReactDOM from "react-dom";

class App extends React.Component<{required: mixed}, *> {
}
const App2: any = App;
const App3: React.ComponentType<{required: mixed}> = App2; // Correctly gives type error here

ReactDOM.render(
  <App3 />,
  document.getElementById("app")
);
```
